subsys/pm/device_runtime.c

```diff
@@ -94,9 +94,20 @@ static void runtime_suspend_work(struct k_work *work)
 	k_condvar_broadcast(&pm->condvar);
 	k_mutex_unlock(&pm->lock);
 
+	/*
+	 * On async put, we have to suspend the domain when the device
+	 * finishes its operation
+	 */
+	if (pm->domain != NULL) {
+		(void)pm_device_runtime_put_async(pm->domain);
+	}
+
 	__ASSERT(ret == 0, "Could not suspend device (%d)", ret);
 }
 
+#define PM_CHECK_RETURN(_ret) \
+	((ret == 0) || (ret == -ENOTSUP) || (ret == -ENOSYS))
```

Comment:
are all these error codes meaningfull for domains? `-ENOSYS` surely not, `-ENOTSUP` probably not if we require domains to implement suspend/resume.

Reply:
That is the thing, power domains is just a regular device and the action callback is not enforced.  Consequently, if a power domain does not implement an action callback, it will never be suspended and will return `-ENOSYS`. If that that makes sense (having a power domain that does not implement action cb) is other question.

Reply:
I think there should be a contract here: a power domain needs to implement suspend and resume. Adding this check will mask implementation issues.

Reply:
Fair enough. Lets enforce domains to implement the action callback and handle SUSPEND/RESUME